Errors during compiling for bk72xx

### The problem

I former working YAML is not compiling anymore with the following errors
```
Compiling .pioenvs/power_meter_1/src/esphome/components/mqtt/custom_mqtt_device.cpp.o
Compiling .pioenvs/power_meter_1/src/esphome/components/mqtt/mqtt_binary_sensor.cpp.o
In file included from src/esphome/components/mqtt/mqtt_client.h:17,
                 from src/esphome/components/mqtt/custom_mqtt_device.h:7,
                 from src/esphome/components/mqtt/custom_mqtt_device.cpp:1:
src/esphome/components/mqtt/mqtt_backend_libretiny.h: In member function 'virtual void esphome::mqtt::MQTTBackendLibreTiny::set_server(esphome::network::IPAddress, uint16_t)':
src/esphome/components/mqtt/mqtt_backend_libretiny.h:23:38: error: invalid 'static_cast' from type 'esphome::network::IPAddress' to type 'uint32_t' {aka 'long unsigned int'}
   23 |     mqtt_client_.setServer(IPAddress(static_cast<uint32_t>(ip)), port);
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from src/esphome/components/mqtt/mqtt_client.h:17,
                 from src/esphome/components/mqtt/mqtt_component.h:11,
                 from src/esphome/components/mqtt/mqtt_binary_sensor.h:7,
                 from src/esphome/components/mqtt/mqtt_binary_sensor.cpp:1:
src/esphome/components/mqtt/mqtt_backend_libretiny.h: In member function 'virtual void esphome::mqtt::MQTTBackendLibreTiny::set_server(esphome::network::IPAddress, uint16_t)':
src/esphome/components/mqtt/mqtt_backend_libretiny.h:23:38: error: invalid 'static_cast' from type 'esphome::network::IPAddress' to type 'uint32_t' {aka 'long unsigned int'}
   23 |     mqtt_client_.setServer(IPAddress(static_cast<uint32_t>(ip)), port);
      |                                      ^~~~~~~~~~~~~~~~~~~~~~~~~
*** [.pioenvs/power_meter_1/src/esphome/components/mqtt/custom_mqtt_device.cpp.o] Error 1
*** [.pioenvs/power_meter_1/src/esphome/components/mqtt/mqtt_binary_sensor.cpp.o] Error 1
```

### Which version of ESPHome has the issue?

ESPHome 2023.10.1

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

Home Assistant 2023.10.3 Supervisor 2023.10.0 Operating System 11.0

### What platform are you using?

ESP8266

### Board

bk72xx

### Component causing the issue

MQTT

### Example YAML snippet

```yaml
esphome:
  name: power_meter_1

bk72xx:
  board: generic-bk7231n-qfn32-tuya

logger:

web_server:

captive_portal:

mdns:

api:
  password: ""

ota:
  password: ""

wifi:
  networks:
  
  - ssid: "xxxxxxxxxxxxx"
    password: "xxxxxxxxxxxxx"
  - ssid: "xxxxxxxxxxxxx"
    password: "xxxxxxxxxxxx"
  # Remove or comment out use_address
  #use_address: upk2esphome-bk7231n.local
  power_save_mode: none

mqtt:
  broker: xxxxxxxxxxx
  username: xxxxx
  password: xxxxxxxxx
  port: 1883
  

uart:
  rx_pin: RX1
  tx_pin: TX1
  baud_rate: 9600

tuya:

sensor:
  - platform: "tuya"
    name: "Voltage"
    sensor_datapoint: 112
    device_class: "voltage"
    state_class: "measurement"
    filters:
    - multiply: 0.1
    - lambda: return x / 1.004;

  - platform: "tuya"
    name: "FREQ"
    sensor_datapoint: 111
    device_class: "frequency"
    state_class: "measurement"
    filters:
    - multiply: 0.01
    - lambda: return x * 1.02;

  - platform: "tuya"
    name: "Current A"
    sensor_datapoint: 113
    state_class: "measurement"
    filters:
    - multiply: 0.01

  - platform: "tuya"
    name: "Current B"
    sensor_datapoint: 114
    filters:
    - multiply: 0.01
  
  - platform: "tuya"
    name: "Forward Energy A"
    sensor_datapoint: 106
    filters:
    - multiply: 0.01

  - platform: "tuya"
    name: "Forward Energy B"
    sensor_datapoint: 108
    filters:
    - multiply: 0.01

  - platform: "tuya"
    name: "Reverse Energy A"
    sensor_datapoint: 107
    filters:
    - multiply: 0.01

  - platform: "tuya"
    name: "Reverse Energy B"
    sensor_datapoint: 109
    filters:
    - multiply: 0.01

  - platform: "tuya"
    name: "Total Power A"
    sensor_datapoint: 101
    filters:
    - multiply: 0.01

  - platform: "tuya"
    name: "Total Power B"
    sensor_datapoint: 105
    filters:
    - multiply: 0.01

binary_sensor:
  - platform: "tuya"
    name: "Direction A"
    sensor_datapoint: 102

  - platform: "tuya"
    name: "Direction B"
    sensor_datapoint: 104

switch:
  - platform: restart
    name: "Meter Restart"
```


### Anything in the logs that might be useful for us?

_No response_

### Additional information

_No response_

--- a/esphome/components/mqtt/mqtt_backend_libretiny.h
+++ b/esphome/components/mqtt/mqtt_backend_libretiny.h
@@ -20,7 +20,7 @@ class MQTTBackendLibreTiny final : public MQTTBackend {
     mqtt_client_.setWill(topic, qos, retain, payload);
   }
   void set_server(network::IPAddress ip, uint16_t port) final {
-    mqtt_client_.setServer(IPAddress(static_cast<uint32_t>(ip)), port);
+    mqtt_client_.setServer(IPAddress(ip), port);
   }
   void set_server(const char *host, uint16_t port) final { mqtt_client_.setServer(host, port); }
 #if ASYNC_TCP_SSL_ENABLED

--- a/esphome/components/mqtt/mqtt_backend_libretiny.h
+++ b/esphome/components/mqtt/mqtt_backend_libretiny.h
@@ -19,9 +19,7 @@ class MQTTBackendLibreTiny final : public MQTTBackend {
   void set_will(const char *topic, uint8_t qos, bool retain, const char *payload) final {
     mqtt_client_.setWill(topic, qos, retain, payload);
   }
-  void set_server(network::IPAddress ip, uint16_t port) final {
-    mqtt_client_.setServer(IPAddress(ip), port);
-  }
+  void set_server(network::IPAddress ip, uint16_t port) final { mqtt_client_.setServer(IPAddress(ip), port); }
   void set_server(const char *host, uint16_t port) final { mqtt_client_.setServer(host, port); }
 #if ASYNC_TCP_SSL_ENABLED
   void set_secure(bool secure) { mqtt_client.setSecure(secure); }

--- a/tests/test9.yaml
+++ b/tests/test9.yaml
@@ -26,3 +26,9 @@ sensor:
     name: ADC
     pin: GPIO23
     update_interval: 1s
+
+mqtt:
+  broker: test.mosquitto.org
+  port: 1883
+  discovery: true
+  discovery_prefix: homeassistant

